File: hackerrank/problem solving/algorithms/2 - implementation/34 - cut the sticks/cut_the_sticks.py
# https://www.hackerrank.com/challenges/cut-the-sticks/problem?isFullScreen=true

def main():
    input()
    arr = read_numbers()
    arr.sort()
    print(*cut_the_sticks(arr), sep='\n')


def read_numbers():
    return list(map(int, input().split()))


# n: length of arr
# k: number of distinct values in arr
# k <= n
# T: O(n)
#    Sorting arr beforehand is O(n log n)
# S: O(k) = O(n) extra space
#    Sorting arr beforehand is O(n) extra space
def cut_the_sticks(arr):
    res = []
    slow, n = 0, len(arr)

    for fast in range(n):
        if arr[slow] == arr[fast]:
            continue
        res.append(n - slow)
        slow = fast
    res.append(n - slow)

    return res


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/34 - cut the sticks/cut_the_sticks.js
// https://www.hackerrank.com/challenges/cut-the-sticks/problem?isFullScreen=true

function main() {
  readLine();
  const arr = readNumbers();
  arr.sort((a, b) => a - b);
  for (const x of cutTheSticks(arr)) console.log(x);
}

function readNumbers() {
  return readLine()
    .split(' ')
    .map((x) => +x);
}

// n: length of arr
// k: number of distinct values in arr
// k <= n
// T: O(n)
//    Sorting arr beforehand is O(n log n)
// S: O(k) = O(n) extra space
//    Sorting arr beforehand is O(n) extra space
function cutTheSticks(arr) {
  const res = [];
  let slow = 0;
  const n = arr.length;

  for (let fast = 0; fast < n; fast++) {
    if (arr[slow] === arr[fast]) continue;
    res.push(n - slow);
    slow = fast;
  }

  res.push(n - slow);
  return res;
}

////////////////////////////////////////////////

const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
let inputLines = [];
rl.on('line', (line) => inputLines.push(line));
rl.on('close', main);

function readLine() {
  return inputLines.shift();
}
